File: modulo04-agentes-autonomos/agent-memory-java/src/test/java/com/iadeva/memory/MemoryTest.java
package com.iadeva.memory;

import com.iadeva.memory.memory.ContextualMemory;
import com.iadeva.memory.memory.EpisodicMemory;
import com.iadeva.memory.memory.LongTermMemory;
import com.iadeva.memory.model.Episode;
import com.iadeva.memory.model.MemoryFact;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.mockito.Mockito;
import org.springframework.ai.embedding.EmbeddingModel;

import java.io.File;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

/**
 * Testes unitários para os 3 tipos de memória persistida.
 * Usa diretório temporário para isolar cada teste — sem efeitos colaterais.
 */
class MemoryTest {

    @TempDir
    File tempDir;

    private LongTermMemory longTermMemory;
    private EpisodicMemory episodicMemory;
    private ContextualMemory contextualMemory;
    private EmbeddingModel embeddingModelMock;

    @BeforeEach
    void setUp() {
        String path = tempDir.getAbsolutePath();

        longTermMemory = new LongTermMemory(path);
        longTermMemory.init();

        episodicMemory = new EpisodicMemory(path);
        episodicMemory.init();

        // Mock do EmbeddingModel — evita chamadas reais à API da OpenAI nos testes
        embeddingModelMock = Mockito.mock(EmbeddingModel.class);
    }

    // -------------------------------------------------------------------------
    // CENÁRIO 1: LongTermMemory não deve duplicar fatos com mesmo content
    // -------------------------------------------------------------------------

    @Test
    @DisplayName("LongTermMemory: não deve duplicar fatos com mesmo content")
    void longTermMemory_naoDeveDuplicar_fatosComMesmoContent() {
        MemoryFact fato = new MemoryFact(
                UUID.randomUUID().toString(),
                "O usuário prefere respostas em português",
                "usuario",
                Instant.now(),
                null
        );

        // Adiciona o mesmo fato duas vezes (mesmo content, id diferente)
        longTermMemory.addFact(fato);
        longTermMemory.addFact(new MemoryFact(
                UUID.randomUUID().toString(), // id diferente
                "O usuário prefere respostas em português", // mesmo content
                "sistema",
                Instant.now(),
                null
        ));

        List<MemoryFact> fatos = longTermMemory.getFacts();
        assertThat(fatos).hasSize(1);
        assertThat(fatos.get(0).content()).isEqualTo("O usuário prefere respostas em português");
    }

    // -------------------------------------------------------------------------
    // CENÁRIO 2: ContextualMemory retorna resultados acima do threshold e ignora abaixo
    // -------------------------------------------------------------------------

    @Test
    @DisplayName("ContextualMemory: retorna resultados acima do threshold e ignora abaixo")
    void contextualMemory_filtraPorThreshold() {
        // Vetores controlados: alta similaridade (cosseno ~1.0) e baixa (~0.0)
        float[] vetorBase = {1.0f, 0.0f, 0.0f};
        float[] vetorSimilar = {0.95f, 0.1f, 0.0f}; // cosseno alto
        float[] vetorDistante = {0.0f, 0.0f, 1.0f};  // cosseno ~ 0.0

        // Mock: primeira chamada = query, segunda = vetorSimilar, terceira = vetorDistante
        when(embeddingModelMock.embed(anyString()))
                .thenReturn(toDoubleList(vetorBase))    // embed da query
                .thenReturn(toDoubleList(vetorSimilar)) // store conteúdo 1
                .thenReturn(toDoubleList(vetorDistante)); // store conteúdo 2

        // Threshold 0.7 — só vetorSimilar deve passar
        ContextualMemory memory = new ContextualMemory(embeddingModelMock, 0.7);

        // Simula stores já realizados (precisamos que o mock forneça embeddings ao store)
        // Estratégia: resetar mock e configurar para store → search
        Mockito.reset(embeddingModelMock);
        when(embeddingModelMock.embed("conteúdo relevante")).thenReturn(toDoubleList(vetorSimilar));
        when(embeddingModelMock.embed("conteúdo irrelevante")).thenReturn(toDoubleList(vetorDistante));
        when(embeddingModelMock.embed("query de busca")).thenReturn(toDoubleList(vetorBase));

        memory.store("conteúdo relevante", Map.of("tipo", "teste"));
        memory.store("conteúdo irrelevante", Map.of("tipo", "teste"));

        List<ContextualMemory.EmbeddingEntry> resultados = memory.search("query de busca", 10);

        // Apenas o conteúdo com alta similaridade deve retornar
        assertThat(resultados).hasSize(1);
        assertThat(resultados.get(0).content()).isEqualTo("conteúdo relevante");
    }

    // -------------------------------------------------------------------------
    // CENÁRIO 3: EpisodicMemory getRecentEpisodes respeita limit
    // -------------------------------------------------------------------------

    @Test
    @DisplayName("EpisodicMemory: getRecentEpisodes deve respeitar o limit informado")
    void episodicMemory_getRecentEpisodes_respeitaLimit() {
        // Adiciona 5 episódios com timestamps distintos
        for (int i = 1; i <= 5; i++) {
            episodicMemory.addEpisode(new Episode(
                    UUID.randomUUID().toString(),
                    "input " + i,
                    List.of("passo 1", "passo 2"),
                    "resultado " + i,
                    List.of(),
                    Instant.now().plusSeconds(i) // timestamps crescentes
            ));
        }

        List<Episode> recentes = episodicMemory.getRecentEpisodes(3);

        assertThat(recentes).hasSize(3);
        // Verifica que os mais recentes vêm primeiro (maior timestamp)
        assertThat(recentes.get(0).input()).isEqualTo("input 5");
        assertThat(recentes.get(1).input()).isEqualTo("input 4");
        assertThat(recentes.get(2).input()).isEqualTo("input 3");
    }

    // Utilitário: converte float[] para List<Double> (formato retornado pelo EmbeddingModel)
    private List<Double> toDoubleList(float[] arr) {
        List<Double> list = new java.util.ArrayList<>();
        for (float v : arr) list.add((double) v);
        return list;
    }
}

File: modulo04-agentes-autonomos/agent-memory-go/memory/memory_test.go
package memory

import (
	"context"
	"strconv"
	"testing"
	"time"

	"agent-memory/model"

	"github.com/google/uuid"
)

func TestLongTermMemory_DoesNotDuplicateFactsWithSameContent(t *testing.T) {
	ltm := NewLongTermMemory(t.TempDir())

	ltm.AddFact(model.MemoryFact{
		ID: uuid.NewString(), Content: "O usuário prefere respostas em português",
		Source: "usuario", ConfirmedAt: time.Now(),
	})
	ltm.AddFact(model.MemoryFact{
		ID: uuid.NewString(), Content: "O usuário prefere respostas em português",
		Source: "sistema", ConfirmedAt: time.Now(),
	})

	facts := ltm.GetFacts()
	if len(facts) != 1 {
		t.Fatalf("esperava 1 fato (deduplicado), obteve %d", len(facts))
	}
	if facts[0].Content != "O usuário prefere respostas em português" {
		t.Errorf("conteudo inesperado: %q", facts[0].Content)
	}
}

type fakeEmbedder struct {
	vectors map[string][]float64
}

func (f *fakeEmbedder) Embed(ctx context.Context, text string) ([]float64, error) {
	return f.vectors[text], nil
}

func TestContextualMemory_FiltersByThreshold(t *testing.T) {
	vetorBase := []float64{1.0, 0.0, 0.0}
	vetorSimilar := []float64{0.95, 0.1, 0.0} // cosseno alto
	vetorDistante := []float64{0.0, 0.0, 1.0} // cosseno ~0.0

	embedder := &fakeEmbedder{vectors: map[string][]float64{
		"conteudo relevante":   vetorSimilar,
		"conteudo irrelevante": vetorDistante,
		"query de busca":       vetorBase,
	}}

	mem := &ContextualMemory{Embedder: embedder, SimilarityThreshold: 0.7}

	ctx := context.Background()
	mem.Store(ctx, "conteudo relevante", map[string]any{"tipo": "teste"})
	mem.Store(ctx, "conteudo irrelevante", map[string]any{"tipo": "teste"})

	results, err := mem.Search(ctx, "query de busca", 10)
	if err != nil {
		t.Fatalf("erro inesperado: %v", err)
	}
	if len(results) != 1 {
		t.Fatalf("esperava 1 resultado acima do threshold, obteve %d", len(results))
	}
	if results[0].Content != "conteudo relevante" {
		t.Errorf("conteudo inesperado: %q", results[0].Content)
	}
}

func TestEpisodicMemory_GetRecentEpisodesRespectsLimit(t *testing.T) {
	em := NewEpisodicMemory(t.TempDir())

	base := time.Now()
	for i := 1; i <= 5; i++ {
		em.AddEpisode(model.Episode{
			ID: uuid.NewString(), Input: "input " + strconv.Itoa(i),
			Steps: []string{"passo 1", "passo 2"}, Outcome: "resultado " + strconv.Itoa(i),
			Timestamp: base.Add(time.Duration(i) * time.Second),
		})
	}

	recent := em.GetRecentEpisodes(3)

	if len(recent) != 3 {
		t.Fatalf("esperava 3 episodios, obteve %d", len(recent))
	}
	if recent[0].Input != "input 5" || recent[1].Input != "input 4" || recent[2].Input != "input 3" {
		t.Errorf("ordem inesperada: %v", []string{recent[0].Input, recent[1].Input, recent[2].Input})
	}
}
